core-to-plc/src/Language/Plutus/Lift/Instances.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE KindSignatures    #-}
+{-# LANGUAGE PolyKinds         #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Language.Plutus.Lift.Instances () where
+
+import qualified Language.PlutusCore             as PLC
+
+import           Language.Plutus.CoreToPLC.Utils
+import           Language.Plutus.Lift.LiftPir
+import           Language.PlutusIR
+
+import qualified Data.ByteString.Lazy            as BSL
+import           Data.Proxy
+
+-- Derived instances
+
+-- This instance ensures that we can apply typeable type constructors to typeable arguments and get a typeable
+-- type. We need the kind variable, so that partial application of type constructors works.
+instance (TypeablePir (f :: * -> k), TypeablePir (a :: *)) => TypeablePir (f a) where
+    typeRep _ = TyApp () <$> typeRep (undefined :: Proxy f) <*> typeRep (undefined :: Proxy a)
+
+instance (TypeablePir (a :: *), TypeablePir (b :: *)) => TypeablePir (a -> b) where
+    typeRep _ = TyFun () <$> typeRep (undefined :: Proxy a) <*> typeRep (undefined :: Proxy b)
+
+-- Primitives
+
+instance TypeablePir Int where
+    typeRep _ = pure $ appSize haskellIntSize (TyBuiltin () PLC.TyInteger)
+
+instance LiftPir Int where
+    lift i = pure $ Constant () $ PLC.BuiltinInt () haskellIntSize $ fromIntegral i
+
+instance TypeablePir BSL.ByteString where
+    typeRep _ = pure $ appSize haskellBSSize (TyBuiltin () PLC.TyByteString)
+
+instance LiftPir BSL.ByteString where
+    lift bs = pure $ Constant () $ PLC.BuiltinBS () haskellBSSize bs
+
```

Comment:
I understand that this PR is mostly concerned with the LiftPlc stuff, but in general would it be possible to define 

```
import           Language.Haskell.TH (Exp, Q)
instance LiftPir (Q Exp) where
...
```
? And if so, can that be used to implement `Language.Plutus.TH.plutus` ?

I suppose it would mean that `lift`ing can fail if something isn't in scope

Reply:
I'm not sure that's a good idea. I don't want to confuse the output of the plugin with the result of lifting things.

I also don't think it's possible - I think TH will complain because it can't splice in a parameter.